### ISSUE TITLE
Usage in case of remote calls with timeout

### ISSUE THREAD

**Issue by PierreR** [yamadapc/haskell-ascii-progress#21]
Hi,

I have been attempting to use `ascii-progress` as a progress bar for remote calls.

It is not possible to know how much time the call would last. I just know that the command would timeout after x sec.

Here is the code so far:

```
main :: IO ()
main = displayConsoleRegions $ do
    pg <- newProgressBar def { pgTotal = 60
                             , pgFormat = "Waiting [:bar], timeout in:eta sec"
                             , pgOnCompletion = Just "After:elapsed sec"
                             }
    race (loop pg) cmd >>= \case
      Left () -> putText "Exit with timeout"
      Right () -> do
	# putText "break it"
        complete pg
	# putText "will never be displayed"

  where
    loop pg = do
        b <- isComplete pg
        unless b $ do
            threadDelay $ 1000 * 1000
            tickN pg 1
            loop pg
    cmd = do
      (_, _, _, b) <- createProcessConcurrent $ shell ("sleep 10; echo result")
      void $ waitForProcessConcurrent b

```
It works but it raises a couple of questions:

- `pgOnCompletion` message needs to make sense when the timeout expires or when a result is displayed. 
- The API is fragile. See `putText` comments in the code. 
- It need `System.Console.Concurrent` (both createProcessConcurrent & waitForProcessConcurrent). Otherwise the progress bar is not even displayed. As `System.Console.Concurrent` is not re-exported in `ascii-progress`, I doubt the code is idiomatic enough.

@joeyh maybe I should use `concurrent-output` alone for this case ?
 
Thanks for your advices.

**Comment by yamadapc** [comment]
This works https://gist.github.com/7d6e46db4c4e1f3ae31324366e520101

- - -
But, where is `putText` from? What is that?

- - -

I'd seem to me your issues start if you want the output from external processes to be handled in some way; that's outside of the scope of this package, but `concurrent-output` should have the primitives for you to do it.

- - -

I could expose more of the interface to run the thread that renders the progress-bar on its own region/to a string and that receives the ticks and completion to update the state so you can manage it.

**Comment by joeyh** [comment]
Yes, thanks to Pedro's use of concurrent-output in ascii-progress, you
can use createProcessConcurrent to make all output of the process(es)
scroll up above the progress bars. I think all you need is to replace
createProcess in his ghist with createProcessConcurrent to get that.

If you follow the link from the newProgressBar API documentation to
displayConsoleRegions, it mentions createProcessConcurrent there.
So it is documented, but there's always room for documentation to be
improved somehow if you have ideas.

-- 
see shy jo

**Comment by PierreR** [comment]
I have updated the gist to use `craeteProcessConcurrent`. If I not mistaken I also need `waitForProcessConcurrent` ?

@joeyh Is there something wrong in my code that would explain the erratic output I see whenever I add  [putStrLn "Command ended first"](https://gist.github.com/yamadapc/7d6e46db4c4e1f3ae31324366e520101#file-test-hs-L25) ?

I feel I am just missing a tiny bit to get this right ;-)

**Comment by joeyh** [comment]
Pi3r wrote:
> I have updated the gist to use craeteProcessConcurrent. If I not mistaken I
> also need waitForProcessConcurrent ?

Yes, it's the only thing that can wait on a ConcurrentProcessHandle.

> @joeyh Is there something wrong in my code that would explain the erratic
> output I see whenever I add putStrLn "Command ended first" ?
> 
> I feel I am just missing a tiny bit to get this right ;-)

By using putStrLn, you bypass all the concurrent-safe output, and
scribble wherever concurrent-output happens to have left the cursor,
which also will confuse it about where the cursor is when it
subsequently refreshes the display. Using outputConcurrent will avoid
these problems.

-- 
see shy jo

**Comment by PierreR** [comment]
>  Using outputConcurrent will avoid these problems.

@joeyh if I just replace putStrLn with `outputConcurrent`, there is not output at all. Maybe an artifact of the `complete` call that would have "closed" the region somehow ?

@yamadapc As soon as this code works, I would PR it as an example if it is OK by you ?

**Comment by PierreR** [comment]
@joeyh Oh that's amazing. It does work if I don't forget to supply `\n` at the end of the string ... I knew I had to supply it but I am surprised by the fact that the whole string won't be output when I omit it.

Now that I think about it a bit more, this is probably due to the `LineBuffering`, so it should be expected ;-)

I will try to integrate all this with `turtle` now.

Thanks for your help !

**Comment by joeyh** [comment]
Pi3r wrote:
> @joeyh Oh that's amazing. It does work if I don't forget to supply \n at the
> end of the string ... I knew I had to supply it but I am surprised by the fact
> that the whole string won't be output when I omit it.

It buffers until newline in this case. Good feedback, I'll improve the
haddocks.

-- 
see shy jo

**Comment by PierreR** [comment]
@joeyh I need to pass some output (the `s` argument) to stdin while using `createProcessConcurrent`:
Here is my try at it:
```
shell' :: Text -> Text -> IO ExitCode
shell' cmd s = do
  let open = do
        (Just inh, Nothing, Nothing, ph) <- createProcessConcurrent $ (Process.shell (toS cmd))
                                                                          { Process.std_in  = Process.CreatePipe
                                                                          , Process.std_out = Process.Inherit
                                                                          , Process.std_err = Process.Inherit
                                                                          }
        return (inh, ph)
      handle (inh, ph) = do
        Text.hPutStrLn inh s
        waitForProcessConcurrent ph
      close (inh, ConcurrentProcessHandle ph) = do
        IO.hClose inh
        Process.terminateProcess ph
  bracket open close handle
```
But the `ConcurrentProcessHandle` data constructor is not exposed. Don't I need to call `terminateProcess` in this example ?

**Comment by PierreR** [comment]
@yamadapc There is something wicked going on ... Whenever I use a bigger timeout (let say `240`):
```
    pg <- newProgressBar def { pgTotal = if doTimeout then 2 else 240
```
After 10 sec or so,  the progress bar would suddenly increase the `:eta` value. For instance the display would go from ` timeout in 223 sec` to `timeout in 320 sec`. 

Here is another visual clue:
```
λ pi3r/haskell/sandbox → time ~/.local/bin/template 
Waiting [===                                             ], timeout in 229.3 sec

λ pi3r/haskell/sandbox → time ~/.local/bin/template 
Waiting [======                                          ], timeout in 251.3 sec

λ pi3r/haskell/sandbox → time ~/.local/bin/template 
Timeout.
After277.5 sec
~/.local/bin/template  2.69s user 0.11s system 1% cpu 4:38.50 total
```

It seems to go up only once sometimes after 5, 10 or 20 sec. At the end the process does take longer than expected.

This happens at every run (even with lower values such as 120 or 60).  Hopefully it should be easy to reproduce on your side. Shall I open a new issue ?

**Comment by joeyh** [comment]
Pi3r wrote:
> But the ConcurrentProcessHandle data constructor is not exposed. Don't I need
> to call terminateProcess in this example ?

I'm not sure what the terminateProcess there would be doing. Cleaning up
after an async exception?

There is not currently a way to force termination of a
ConcurrentProcessHandle, and I agree that's a bit of an oversight since
sometimes a signal really does need to be sent to a process.

The reason that the ConcurrentProcessHandle constructor is not exported
is that would allow this code to be written:

	(_, _, _, p) <- createProcess (proc "foo" [])
	waitForProcessConcurrent (ConcurrentProcessHandle p)

Which would either block for longer than expected, or potentially
misbehave in some other way, since it's violating an expected invariant
that the process was started with createProcessConcurrent.

Perhaps the thing to do is provide:

	withProcessHandle :: ConcurrentProcessHandle -> (ProcessHandle -> IO a) -> IO a

It should be fine to run any function like terminateProcess,
interruptProcessGroupOf with the ProcessHandle
contained within a ConcurrentProcessHandle.

But, that could be misused thus:

	(_, _, _, p) <- createProcessConcurrent (proc "foo" [])
	withProcessHandle p waitForProcess

That *might* be ok, but i am not 100% sure. It might result in a
resource leak. If it is ok, then it would follow there's no reason for
waitForProcessConcurrent to exist, and thus no need for
ConcurrentProcessHandle!

(it may be that waitForProcessConcurrent is unncessary complexity caused
by working around the now-fixed
https://github.com/haskell/process/issues/46)

-- 
see shy jo

**Comment by joeyh** [comment]
In fact, I did some digging and found the original docs for 
waitForProcessConcurrent:

-- This is necessary because `System.Process.waitForProcess` has a
-- race condition when two threads check the same process. If the race
-- is triggered, one thread will successfully wait, but the other
-- throws a DoesNotExist exception.

So, I do think that concurrent-process could be simplified now,
if it was made to depend on the version of process that fixed that race.

-- 
see shy jo

**Comment by joeyh** [comment]
Indeed, it simplifies nicely now. Released concurrent-output 1.10.0
with ConcurrentProcessHandle = ProcessHandle.

-- 
see shy jo

**Comment by PierreR** [comment]
@joeyh Thanks I will test the new version asap. 

I have created a new issue for the `:eta` wierd problem that I have described here (#22)
